lib/realtime/vehicle.ex

```diff
@@ -214,6 +225,49 @@ defmodule Realtime.Vehicle do
     }
   end
 
+  @doc """
+  Does this vehicle have a trip assignment from Busloc, but not from Swiftly?
+  That is a sign that Swiftly thinks the vehicle is off course, or not on any
+  trip for some other reason.
+  """
+  @spec off_course?([DataDiscrepancy.t()] | DataDiscrepancy.t()) :: boolean
+  def off_course?(data_discrepancies) when is_list(data_discrepancies) do
+    trip_id_discrepency =
+      Enum.find(data_discrepancies, fn data_discrepancy ->
+        data_discrepancy.attribute == :trip_id
+      end)
+
+    off_course?(trip_id_discrepency)
+  end
+
+  def off_course?(nil), do: false
+
+  def off_course?(%{sources: sources}) do
+    case Enum.find(sources, fn source -> source.id == "swiftly" end) do
+      %{value: nil} ->
+        true
+
+      _ ->
+        false
+    end
+  end
+
+  @doc """
+  Check whether the vehicle is off course. If so, check if the assigned block
+  was scheduled to end over an hour ago. We give the buffer so that we don't
+  tag a bus that is on a detour and late, and should thus still be shown on the
+  route ladder.
+  """
+  @spec active_block?(boolean(), Block.t(), Util.Time.timestamp()) :: boolean()
+  def active_block?(false, _block, _now), do: true
+
+  def active_block?(true, block, now) do
```

Comment:
Unnamed boolean parameters can be pretty mysterious and easy to call incorrectly.

That parameter may not even be necessary. I think `active_block?` makes sense even for on course buses. This field could get populated for all buses and then the frontend would check `vehicle.isOffCourse && !vehicle.activeBlock`.

(Even if you don't end up using `is_off_course` to calculate `active_block`, I still like having the off course calculations on the backend.)

Reply:
Good point. I think it's important to keep all that logic in a single place, but I've added a name to the parameter for more clarity.